### ISSUE TITLE
Local to global resource mapping uniqueness with static const variable

### ISSUE THREAD

**Issue by onovytskyi** [microsoft/DirectXShaderCompiler#2529]
Hello,

I have code like this (original code is much bigger, this is stripped out version)
```
struct InStruct
{
    float Pos : SV_Position;
};

struct OutStruct
{
    float EdgeTess[3] : SV_TessFactor;
};

[domain("tri")]
[partitioning("fractional_odd")]
[outputtopology("triangle_cw")]
[outputcontrolpoints(3)]
[patchconstantfunc("PatchConsts")]
InStruct main(InputPatch<InStruct, 3> In, uint PointID : SV_OutputControlPointID, uint PatchID : SV_PrimitiveID)
{
    InStruct Out;
    Out = In[PointID];
    return Out;
}

SamplerState sampler1 : register(s0);
Texture2D tex1 : register(t0);

struct ResourceStruct
{
    Texture2D tex;
};

ResourceStruct CreateResources()
{
    ResourceStruct i;
    i.tex = tex1;
    return i;
};

static const ResourceStruct resources = CreateResources();

OutStruct PatchConsts(InputPatch<InStruct, 3> In, uint PatchID : SV_PrimitiveID)
{
    OutStruct Out = (OutStruct)0;

    // next line will trigger error: local resource not guaranteed to map to unique global resource. 
    Out.EdgeTess[0] = resources.tex.SampleLevel(sampler1, float2(0, 0), 0).x;

    return Out;
}
```
which returns aforementioned error on compilation. Right now I only managed to repro this using HS. Similar constructs in other shader types seems working fine.
I believe this construct must work, correct me if I'm wrong.

Thanks,
Oleksandr

**Comment by tex3d** [comment]
Looks like a bug.  Thanks for the clear repro.  Seems like a work around is possible, but effort required depends on your code base or shader generator.  We have some issues related to function-based global initializers, but our priority isn't high for these at the moment.

**Comment by pow2clk** [comment]
Fixed by #3122
